analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ConditionEvaluatesToConstantBase.cs

```diff
@@ -63,21 +75,103 @@ public override void ExecutionCompleted()
         var alwaysTrue = trueOperations.Except(falseOperations);
         var alwaysFalse = falseOperations.Except(trueOperations);
 
-        foreach (var operation in alwaysTrue)
+        foreach (var (operation, block) in alwaysTrue.Select(x => (x.Key, x.Value)))
         {
-            ReportIssue(operation, true);
+            ReportIssue(operation, block, true);
         }
-        foreach (var operation in alwaysFalse)
+        foreach (var (operation, block) in alwaysFalse.Select(x => (x.Key, x.Value)))
         {
-            ReportIssue(operation, false);
+            ReportIssue(operation, block, false);
         }
 
         base.ExecutionCompleted();
     }
 
-    private void ReportIssue(IOperation operation, bool conditionEvaluation)
+    private void ReportIssue(IOperation operation, BasicBlock block, bool conditionIsTrue)
+    {
+        var issueMessage = operation.Kind == OperationKindEx.IsNull ? MessageNullCoalescing : string.Format(MessageBool, conditionIsTrue);
+        var secondaryLocations = SecondaryLocations(block, conditionIsTrue);
+        if (secondaryLocations.Any())
+        {
+            ReportIssue(Rule2583, operation, secondaryLocations, string.Format(MessageUnreachable, issueMessage));
+        }
+        else
+        {
+            ReportIssue(Rule2589, operation, null, string.Format(issueMessage, conditionIsTrue));
+        }
+    }
+
+    private List<Location> SecondaryLocations(BasicBlock block, bool conditionIsTrue)
+    {
+        List<Location> locations = new();
+        IOperation currentStart = null;
+        IOperation currentEnd = null;
+        var unreachable = UnreachableOperations(block, conditionIsTrue).Where(x => !IsIgnoredLocation(x.Syntax));
+        foreach (var operation in unreachable.Concat(reached).OrderBy(x => x.Syntax.SpanStart))
+        {
+            if (unreachable.Contains(operation))
```

Comment:
This `Contains` is on `IEnumerable` with `Where`, making it very expensive. `ToHashSet` is missing.